docs/irssi.1
.TH Irssi 1 "June 2015" "Irssi IRC client"
.SH NAME
Irssi \- a modular IRC client for UNIX
.SH SYNOPSIS
.B irssi
[--config=PATH] [--home=PATH] [-dv!?] [-c server] [-p port] [-n nickname]
[-w password] [-h hostname]
.SH DESCRIPTION
.B Irssi
is a modular Internet Relay Chat client; it is highly extensible and
very secure. Being a fullscreen, termcap based client with many
features,
.B Irssi
is easily extensible through scripts and modules.
.SH OPTIONS
.TP
.BI "\-\-config="FILE
use 
.I FILE
instead of ~/.irssi/config
.TP
.BI "\-\-home="PATH
.I PATH 
specifies the home directory of Irssi; default is 
.BR ~/.irssi
.TP
.BI "\-c, \-\-connect="SERVER
connect to 
.I SERVER
.TP
.BI "\-w, \-\-password="PASSWORD
use
.I PASSWORD 
for authentication.
.TP
.BI "\-p, \-\-port="PORT
automatically connect to 
.I PORT 
on server.
.TP
.BI "\-!, \-\-noconnect"
disable autoconnecting of servers
.TP
.BI "\-n, \-\-nick="NICKNAME
specify 
.I NICKNAME 
as your nick
.TP
.BI "\-h, \-\-hostname="HOSTNAME
use
.I HOSTNAME
for your irc session
.TP
.BI "\-v, \-\-version"
display the version of Irssi
.TP
.BI "\-?, \-\-help"
show a help message
.SH SEE ALSO
.B Irssi
has a solid amount of documentation available; check /HELP or look online
at https://irssi.org
.SH FILES
.TP
.I ~/.irssi/config
personal configuration file
.TP
.I ~/.irssi/config.autosave
automatic save of the personal config file when it was changed externally
.TP
.I ~/.irssi/default.theme
default irssi theme
.TP
.I ~/.irssi/away.log
logged messages in away status
.TP
.I ~/.irssi/scripts/
default scripts directory
.TP
.I ~/.irssi/scripts/autorun/
directory containing links to scripts that should be loaded
automatically on startup
.TP
.I ~/.irssi/startup
file containing a list of commands to execute on startup
.SH AUTHORS/CREDITS
.B Irssi
was written by Timo Sirainen; this manpage was written by Istvan Sebestyen, Stefan Tomanek, Geert Hauwaerts.


NEWS
v1.2-head 2018-xx-xx  The Irssi team <staff@irssi.org>

v1.1.0 2018-01-15 The Irssi team <staff@irssi.org>
	* Colour is now re-set when reaching a comma, matching mIRC
	  behaviour (#742, #740, #790)
	* Irssi now shows the initial nick and name on first start
	  (#785, #786)
	* lynx is no longer required to run autogen.sh (#81, #781)
	* The command history no longer permits wrapping around (#686)
	* /foreach now correctly sends arguments as commands, stopping
	  you from embarassing AMSGs (#659)
	* /server does not connect to servers anymore, use /server
	  connect to change servers (#559, #649).
	* The net_ip_compare API function is now deprecated, and the
          previously deprecated net_connect has been removed
          (#770). By Will Storey (#770)
	+ Add an option to ignore all channels or ignore all queries
	  using /set activity_hide_targets. By Jari Matilainen (#612,
	  #779)
	+ Add a startup warning if the TERM var is wrong inside
	  tmux/screen (#726)
	+ Add option to hide certain levels from the textbuffer using
	  /window hidelevel (#746, #808)
	+ Irssi now has its first unit test (for mode parsing). By
	  Will Storey (#793)
	+ Added access to global command history when using window
	  history, and a binding to erase entries from the command
	  history (erase_history_entry) (#762)
	+ -alternate_nick is now available as a network specific
	  property. By Paul Townsend (#120, #771)
	+ On FreeBSD, Irssi now supports Capsicum sandbox (/capsicum
	  enter). By Edward Tomasz Napierala (#735, #755, #772)
	+ Filenames (directories) ending with a / now tab-complete
	  (#741)
	+ UTF-8 should now work in regular expressions when using
	  GRegex (the default) (#636, #653)
	+ Nicks are now properly escaped on completion. By Oscar
	  Linderholm (#693, #709)
	+ /server add -port <num> now works. By Jari Matilainen (#703)
	+ Add a setting key_timeout to make key sequences
	  automatically re-set when not finished (#644, #645)
	+ Warn users about expired client certificates, as servers may
	  refuse them (#211, #627)
	+ Add a new net_start_ssl function for StartTLS. This is
	  available from ABI 8 and can be used by protocol modules
	  (#615, #622).
	+ The %# code is now stored in the textbuffer, so for example
	  web scripts can make use of it (#626)
	+ Add new setting break_wide which can be used to enable
	  breaking of wide characters (for east-asian
	  users). Originally from FreeBSD ports. (#625)
	+ Add fuzzing code (#610, #620, #701, #713)
	- Netsplits show properly again (#812)
	- Do not error on blank lines when using /exec -o. By Fabian
	  Kurz (FS#902, #805)
	- Detect used nickname as reported by server. By Alexandre
	  Morignot (#219, #804)
	- Prevent use after free error during the execution of some
	  commands. Found by Joseph Bisch. (GL#17, GL!24)
	- Fix MODE parameter parsing when colon was used at a place
	  Irssi didn't expect (#601, #766)
	- Fixed code to compile with
	  -Werror=declaration-after-statement (#795)
	- Clang-format is now supported for git-clang-format (#784)
	- Fix use after free when changing the network of
	  hilights. Reported by Rui Mathias. (#787, #788)
	- Fix positioning error when tab-completing non-ascii
	  strings. (#752, #754)
	- In-development issues (#750, #751)
	- Clarify Alis in /help list (#699, #712)
	- Improve /lastlog performance from O(N^2) to O(N) (#715)
	- Fix a segfault on "script destroyed" signal. By Stephen
	  Oberholtzer (#660, #661).
	- Fix early ISON error (#596, #647)
	- Documentation improvements. By Paolo Martini (#639).
	  By Tristan Pepin (#731).  By Paul Townsend (#684, #736).
	  By Will Storey (#777)
	- Minor cleanups (#590). By Edward Tomasz Napierala (#734,
	  #738)
	- Fix space issue in glib-2.0.m4 (#621)

v1.0.6 2018-01-07 The Irssi team <staff@irssi.org>
	- Fix invalid memory access when reading hilight configuration
          (#787, #788).
	- Fix null pointer dereference when the channel topic is set
          without specifying a sender (GL#20, GL!25).
	- Fix return of random memory when using incomplete escape
          codes (GL#21, GL!26).
	- Fix heap buffer overflow when completing certain strings
          (GL#19, GL!27).
	- Fix return of random memory when using an incomplete
          variable argument (GL#18, GL!28).

v1.0.5 2017-10-23  The Irssi team <staff@irssi.org>
	- Fix missing -sasl_method '' in /NETWORK (#718, #719).
	- Fix incorrect restoration of term state when hitting SUSP
          inside screen (#737, #733).
	- Fix out of bounds read when compressing colour
          sequences. Found by Hanno Böck (GL#12, GL!18).
	- Fix use after free condition during a race condition when
          waiting on channel sync during a rejoin (GL#13, GL!19).
	- Fix null pointer dereference when parsing certain malformed
          CTCP DCC messages (GL#14, GL!20).
	- Fix crash due to null pointer dereference when failing to
          split messages due to overlong nick or target (GL#15, GL!21).
	- Fix out of bounds read when trying to skip a safe channel ID
          without verifying that the ID is long enough (GL#16, GL!22).
	- Fix return of random memory when inet_ntop failed (#769).
	- Minor statusbar help update. By Robert Bisewski (#758,
          #763).

v1.0.4 2017-07-07  The Irssi team <staff@irssi.org>
	- Fix null pointer dereference when parsing invalid timestamp (GL#10,
	  GL!15). Reported by Brian 'geeknik' Carpenter.
	- Fix use-after-free condition when removing nicks from the internal
	  nicklist (GL#11, GL!16). Reported by Brian 'geeknik' Carpenter.
	- Fix incorrect string comparison in DCC file names (#714).
	- Fix regression in Irssi 1.0.3 where it would claim "Invalid time '-1'"
	  (#716, #722).
	- Fix a bug when using \n to separate lines with expand_escapes (#723).
	- Retain screen output on improper exit, to better see any error
	  messages (#287, #721).
	- Minor help update (#729).

v1.0.3 2017-06-06  The Irssi team <staff@irssi.org>
	- Fix out of bounds read when scanning expandos (GL!11).

src/fe-text/irssi.c
/*
 irssi.c : irssi

    Copyright (C) 1999-2000 Timo Sirainen

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include "module.h"
#include "module-formats.h"
#include "modules-load.h"
#include "args.h"
#include "signals.h"
#include "levels.h"
#include "core.h"
#include "settings.h"
#include "session.h"
#include "servers.h"

#include "printtext.h"
#include "fe-common-core.h"
#include "fe-settings.h"
#include "themes.h"

#include "term.h"
#include "gui-entry.h"
#include "mainwindows.h"
#include "gui-printtext.h"
#include "gui-readline.h"
#include "statusbar.h"
#include "gui-windows.h"
#include "irssi-version.h"

#include <signal.h>
#include <locale.h>

#ifdef HAVE_STATIC_PERL
void perl_core_init(void);
void perl_core_deinit(void);

void fe_perl_init(void);
void fe_perl_deinit(void);
#endif

void irc_init(void);
void irc_deinit(void);

void fe_common_irc_init(void);
void fe_common_irc_deinit(void);

void gui_expandos_init(void);
void gui_expandos_deinit(void);

void textbuffer_commands_init(void);
void textbuffer_commands_deinit(void);

void lastlog_init(void);
void lastlog_deinit(void);

void mainwindow_activity_init(void);
void mainwindow_activity_deinit(void);

void mainwindows_layout_init(void);
void mainwindows_layout_deinit(void);

static int dirty, full_redraw;

static GMainLoop *main_loop;
int quitting;

static int display_firsttimer = FALSE;
static unsigned int user_settings_changed = 0;


static void sig_exit(void)
{
        quitting = TRUE;
}

static void sig_settings_userinfo_changed(gpointer changedp)
{
	user_settings_changed = GPOINTER_TO_UINT(changedp);
}

/* redraw irssi's screen.. */
void irssi_redraw(void)
{
	dirty = TRUE;
        full_redraw = TRUE;
}

void irssi_set_dirty(void)
{
        dirty = TRUE;
}

static void dirty_check(void)
{
	if (!dirty)
		return;

        term_resize_dirty();

	if (full_redraw) {
                full_redraw = FALSE;

		/* first clear the screen so curses will be
		   forced to redraw the screen */
		term_clear();
		term_refresh(NULL);

		mainwindows_redraw();
		statusbar_redraw(NULL, TRUE);
	}

	mainwindows_redraw_dirty();
        statusbar_redraw_dirty();
	term_refresh(NULL);

        dirty = FALSE;
}

static void textui_init(void)
{
#ifdef SIGTRAP
	struct sigaction act;

	sigemptyset(&act.sa_mask);
	act.sa_flags = 0;
	act.sa_handler = SIG_IGN;
	sigaction(SIGTRAP, &act, NULL);
#endif

	irssi_gui = IRSSI_GUI_TEXT;
	core_init();
	irc_init();
	fe_common_core_init();
	fe_common_irc_init();

	theme_register(gui_text_formats);
	signal_add("settings userinfo changed", (SIGNAL_FUNC) sig_settings_userinfo_changed);
	signal_add_last("gui exit", (SIGNAL_FUNC) sig_exit);
}

static void textui_finish_init(void)
{
	quitting = FALSE;

	term_refresh_freeze();
	textbuffer_init();
	textbuffer_view_init();
	textbuffer_commands_init();
	gui_expandos_init();
	gui_printtext_init();
	gui_readline_init();
	lastlog_init();
	mainwindows_init();
	mainwindow_activity_init();
	mainwindows_layout_init();
	gui_windows_init();
	statusbar_init();
	term_refresh_thaw();

	settings_check();

	module_register("core", "fe-text");

#ifdef HAVE_STATIC_PERL
	perl_core_init();
	fe_perl_init();
#endif

	dirty_check();

	fe_common_core_finish_init();
	signal_emit("irssi init finished", 0);
	statusbar_redraw(NULL, TRUE);

	if (servers == NULL && lookup_servers == NULL) {
		printformat(NULL, NULL, MSGLEVEL_CRAP|MSGLEVEL_NO_ACT, TXT_IRSSI_BANNER);
	}

	if (display_firsttimer) {
		printformat(NULL, NULL, MSGLEVEL_CRAP|MSGLEVEL_NO_ACT, TXT_WELCOME_FIRSTTIME);
	}

	/* see irc-servers-setup.c:init_userinfo */
	if (user_settings_changed)
		printformat(NULL, NULL, MSGLEVEL_CLIENTNOTICE, TXT_WELCOME_INIT_SETTINGS);
	if (user_settings_changed & USER_SETTINGS_REAL_NAME)
		fe_settings_set_print("real_name");
	if (user_settings_changed & USER_SETTINGS_USER_NAME)
		fe_settings_set_print("user_name");
	if (user_settings_changed & USER_SETTINGS_NICK)
		fe_settings_set_print("nick");
	if (user_settings_changed & USER_SETTINGS_HOSTNAME)
		fe_settings_set_print("hostname");

	term_environment_check();
}

static void textui_deinit(void)
{
	signal(SIGINT, SIG_DFL);

        term_refresh_freeze();
	while (modules != NULL)
		module_unload(modules->data);

#ifdef HAVE_STATIC_PERL
        perl_core_deinit();
        fe_perl_deinit();
#endif

	dirty_check(); /* one last time to print any quit messages */
	signal_remove("settings userinfo changed", (SIGNAL_FUNC) sig_settings_userinfo_changed);
	signal_remove("gui exit", (SIGNAL_FUNC) sig_exit);

	lastlog_deinit();
	statusbar_deinit();
	gui_printtext_deinit();
	gui_readline_deinit();
	gui_windows_deinit();
	mainwindows_layout_deinit();
	mainwindow_activity_deinit();
	mainwindows_deinit();
	gui_expandos_deinit();
	textbuffer_commands_deinit();
	textbuffer_view_deinit();
	textbuffer_deinit();

	term_refresh_thaw();
	term_deinit();

	theme_unregister();

	fe_common_irc_deinit();
	fe_common_core_deinit();
	irc_deinit();
	core_deinit();
}

static void check_files(void)
{
	struct stat statbuf;

	if (stat(get_irssi_dir(), &statbuf) != 0) {
		/* ~/.irssi doesn't exist, first time running irssi */
		display_firsttimer = TRUE;
	}
}

int main(int argc, char **argv)
{
	static int version = 0;
	static GOptionEntry options[] = {
		{ "version", 'v', 0, G_OPTION_ARG_NONE, &version, "Display Irssi version", NULL },
		{ NULL }
	};
	int loglev;

	core_register_options();
	fe_common_core_register_options();
	args_register(options);
	args_execute(argc, argv);

 	if (version) {
		printf(PACKAGE_TARNAME" " PACKAGE_VERSION" (%d %04d)\n",
		       IRSSI_VERSION_DATE, IRSSI_VERSION_TIME);
		return 0;
	}

	srand(time(NULL));

	quitting = FALSE;
	core_preinit(argv[0]);

	check_files();

#ifdef HAVE_SOCKS
	SOCKSinit(argv[0]);
#endif

	/* setlocale() must be called at the beginning before any calls that
	   affect it, especially regexps seem to break if they're generated
	   before this call.

	   locales aren't actually used for anything else than autodetection
	   of UTF-8 currently..

	   furthermore to get the users's charset with g_get_charset() properly
	   you have to call setlocale(LC_ALL, "") */
	setlocale(LC_ALL, "");

	loglev = g_log_set_always_fatal(G_LOG_FATAL_MASK | G_LOG_LEVEL_CRITICAL);
	textui_init();

	if (!term_init()) {
		fprintf(stderr, "Can't initialize screen handling.\n");
		return 1;
	}

	g_log_set_always_fatal(loglev);
	textui_finish_init();
	main_loop = g_main_new(TRUE);

	/* Does the same as g_main_run(main_loop), except we
	   can call our dirty-checker after each iteration */
	while (!quitting) {
		if (reload_config) {
			/* SIGHUP received, do /RELOAD */
			reload_config = FALSE;
			signal_emit("command reload", 1, "");
		}

		dirty_check();

		term_refresh_freeze();
		g_main_iteration(TRUE);
		term_refresh_thaw();
	}

	g_main_destroy(main_loop);
	textui_deinit();

	session_upgrade(); /* if we /UPGRADEd, start the new process */
	return 0;
}
